api/alarms/index.js

```diff
@@ -0,0 +1,182 @@
+
+    // Log events in reverse chronological order
+    this.logElement.insertBefore(logLine, this.logElement.firstChild);
+
+    // Log events in chronological order
+    // this.logElement.appendChild(logLine);
+  }
+
+  async handleAlarm(alarm) {
+    let json = JSON.stringify(alarm, null, 2).replace(/\s+/g, ' ');
+    this.logMessage(`Alarm "${alarm.name}" fired\n${json}}`);
+    await this.refreshDisplay();
+  }
+
+  async handleCancelAlarm(event) {
+    if (!event.target.classList.contains('alarm-row__cancel-button')) {
+      return;
+    }
+
+    let name = event.target.parentElement.dataset.name;
+    await this.cancelAlarm(name);
+    await this.refreshDisplay();
+  }
+
+  async cancelAlarm(name) {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve, reject) => {
+      chrome.alarms.clear(name, (wasCleared) => {
+        if (wasCleared) {
+          this.logMessage(`Manager: canceled alarm "${name}"`)
+          resolve(wasCleared);
+        } else {
+          this.logMessage(`Manager: could not canceled alarm "${name}"`)
+          reject(wasCleared);
+        }
+      });
+    });
+  }
+
+  // Thin wrapper around alarms.create to log creation event
+  createAlarm(name, alarmInfo) {
+    chrome.alarms.create(name, alarmInfo);
+    let json = JSON.stringify(alarmInfo, null, 2).replace(/\s+/g, ' ');
+    this.logMessage(`Created "${name}"\n${json}`);
+    this.refreshDisplay();
+  }
+
+  renderAlarm(alarm, isLast) {
+    let alarmEl = document.createElement('div');
+    alarmEl.classList.add('alarm-row');
+    alarmEl.dataset.name = alarm.name;
+    alarmEl.innerText = JSON.stringify(alarm, 0, 2) + (isLast ? '' : ',');
+
+    let cancelButton = document.createElement('button');
+    cancelButton.classList.add('alarm-row__cancel-button');
+    cancelButton.innerText = 'cancel';
+    alarmEl.appendChild(cancelButton);
+
+    this.displayElement.appendChild(alarmEl);
+  }
+
+  async cancelAllAlarms() {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve, reject) => {
+      chrome.alarms.clearAll((wasCleared) => {
+        if (wasCleared) {
+          resolve(wasCleared);
+        } else {
+          reject(wasCleared);
+        }
+      });
+    })
+  }
+
+  async populateDisplay() {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve) => {
+      chrome.alarms.getAll((alarms) => {
+        for (let [index, alarm] of alarms.entries()) {
+          let isLast = index === alarms.length - 1;
+          this.renderAlarm(alarm, isLast);
+        }
+        resolve();
+      });
+    });
+  }
+
+  // Simple locking mechanism to prevent multiple concurrent refreshes
+  #refreshing = false;
+
+  async refreshDisplay() {
+    if (this.#refreshing) return; // refresh in progress, bail
+
+    this.#refreshing = true;      // acquire lock
```

Comment:
this should be

```js
this.#refreshing = true;
try {
  // do await here
} finally {
  this.#refreshing = false;
}
```

Reply:
Thanks. I find try/catch in async functions to be awkward so my brain likes to gloss over it in some cases 😅